File: android/src/main/kotlin/com/royzed/simple_bg_location/data/Position.kt
package com.royzed.simple_bg_location.data

import android.location.Location
import android.os.Build
import com.royzed.simple_bg_location.utils.toIso8601String
import java.util.*

data class Position(
    val uuid: UUID = UUID.randomUUID(),
    val latitude: Double,
    val longitude: Double,
    val timestamp: Long,
    val altitude: Double = -1.0,
    val altitudeAccuracy: Double = -1.0,
    val accuracy: Double = -1.0,
    val heading: Double = -1.0,
    val headingAccuracy: Double = -1.0,
    val floor: Int? = null,
    val speed: Double = -1.0,
    val speedAccuracy: Double = -1.0,
    val isMocked: Boolean = false
) {
    override fun toString(): String {
        val strTime = Date(timestamp).toIso8601String()
        return "$strTime - [$latitude, $longitude], altitude: $altitude"
    }

    fun toMap(): Map<String, Any> {
        val position: MutableMap<String, Any> = mutableMapOf()
        position["latitude"] = latitude
        position["uuid"] = uuid.toString()
        position["longitude"] = longitude
        position["timestamp"] = timestamp
        position["altitude"] = altitude
        position["altitudeAccuracy"] = altitudeAccuracy
        position["accuracy"] = accuracy
        position["heading"] = heading
        position["headingAccuracy"] = headingAccuracy
        if (floor != null) position["floor"] = floor
        position["speed"] = speed
        position["speedAccuracy"] = speedAccuracy
        position["isMocked"] = isMocked

        return position.toMap()
    }

    companion object {
        @JvmStatic
        fun fromLocation(location: Location): Position {
            val altitudeAccuracy: Double =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && location.hasVerticalAccuracy())
                    location.verticalAccuracyMeters.toDouble()
                else -1.0
            val headingAccuracy: Double =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && location.hasBearingAccuracy())
                    location.bearingAccuracyDegrees.toDouble()
                else -1.0
            val speedAccuracy: Double =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && location.hasSpeedAccuracy())
                    location.speedAccuracyMetersPerSecond.toDouble()
                else -1.0
            val isMocked =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
                    location.isMock
                else {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 )
                        location.isFromMockProvider
                    else false
                }

            return Position(
                latitude = location.latitude,
                longitude = location.longitude,
                timestamp = location.time,
                altitude = if (location.hasAltitude()) location.altitude else -1.0,
                altitudeAccuracy = altitudeAccuracy,
                accuracy = if (location.hasAccuracy()) location.accuracy.toDouble() else -1.0,
                heading = location.bearing.toDouble(),
                headingAccuracy = headingAccuracy,
                speed = location.speed.toDouble(),
                speedAccuracy = speedAccuracy,
                isMocked = isMocked
            )
        }
    }
}

File: ios/Classes/domain/Position.swift
//
//  Position.swift
//  simple_bg_location
//
//  Created by Roy Wang on 2022-06-22.
//

import Foundation
import CoreLocation

struct Position {
    let uuid: String
    let latitude: Double
    let longitude: Double
    let timestamp: Int
    let altitude: Double?
    let altitudeAccuracy: Double?
    let accuracy: Double?
    let heading: Double?
    let headingAccuracy: Double?
    let floor: Int?
    let speed: Double?
    let speedAccuracy: Double?
    let isMocked: Bool
    
    func toMap() -> Dictionary<String, Any?> {
        return [
            "uuid": uuid,
            "latitude": latitude,
            "longitude": longitude,
            "timestamp": timestamp,
            "altitude": altitude,
            "altitudeAccuracy": altitudeAccuracy,
            "accuracy": accuracy,
            "heading": heading,
            "headingAccuracy": headingAccuracy,
            "floor": floor,
            "speed": speed,
            "speedAccuracy": speedAccuracy,
            "isMocked": isMocked
        ];
    }
    
    static func fromLocation(_ location: CLLocation?) -> Position? {
        if (location == nil) {
            return nil
        }
        let uuid = UUID()
        var headingAccuracy: Double = -1.0
        if #available(iOS 13.4, *) {
            headingAccuracy = Double(location!.courseAccuracy)
        }
        
        return Position(
            uuid: uuid.uuidString,
            latitude: location!.coordinate.latitude,
            longitude: location!.coordinate.longitude,
            timestamp: Int(location!.timestamp.timeIntervalSince1970)*1000,
            altitude: location!.altitude,
            altitudeAccuracy: Double(location!.verticalAccuracy),
            accuracy: location!.horizontalAccuracy,
            heading: location!.course,
            headingAccuracy: headingAccuracy,
            floor: location!.floor?.level,
            speed: location!.speed,
            speedAccuracy: location!.speedAccuracy,
            isMocked: false
        )
    }
}
